src/RoombaEstimator.cpp

```diff
@@ -0,0 +1,368 @@
+    double pix_R = std::hypot(ph, pw) * 0.5;
+
+    double max_phi = settings_.bottom_camera_aov * M_PI / 360;
+    double pix_focal = pix_R / std::tan(max_phi);
+    double theta = std::atan2(py, px);
+
+    double camera_radius = pix_r / pix_focal;
+
+    ray.vector.x = camera_radius * std::cos(theta);
+    ray.vector.y = camera_radius * std::sin(theta);
+    ray.vector.z = 1;
+
+    double norm = std::sqrt(std::pow(ray.vector.x, 2)
+                          + std::pow(ray.vector.y, 2)
+                          + std::pow(ray.vector.z, 2));
+
+    ray.vector.x /= norm;
+    ray.vector.y /= norm;
+    ray.vector.z /= norm;
+}
+
+void RoombaEstimator::getDynamicSettings(
+        iarc7_vision::RoombaEstimatorConfig& config)
+{
+    if (!dynamic_reconfigure_called_) {
+        config.hsv_slice_h_green_min = settings_.hsv_slice_h_green_min;
+        config.hsv_slice_h_green_max = settings_.hsv_slice_h_green_max;
+        config.hsv_slice_h_red1_min  = settings_.hsv_slice_h_red1_min;
+        config.hsv_slice_h_red1_max  = settings_.hsv_slice_h_red1_max;
+        config.hsv_slice_h_red2_min  = settings_.hsv_slice_h_red2_min;
+        config.hsv_slice_h_red2_max  = settings_.hsv_slice_h_red2_max;
+        config.hsv_slice_s_min       = settings_.hsv_slice_s_min;
+        config.hsv_slice_s_max       = settings_.hsv_slice_s_max;
+        config.hsv_slice_v_min       = settings_.hsv_slice_v_min;
+        config.hsv_slice_v_max       = settings_.hsv_slice_v_max;
+
+        config.min_roomba_blob_size  = settings_.min_roomba_blob_size;
+        config.max_roomba_blob_size  = settings_.max_roomba_blob_size;
+
+        config.morphology_size = settings_.morphology_size;
+        config.morphology_iterations = settings_.morphology_iterations;
+
+        dynamic_reconfigure_called_ = true;
+    } else {
+        settings_.hsv_slice_h_green_min = config.hsv_slice_h_green_min;
+        settings_.hsv_slice_h_green_max = config.hsv_slice_h_green_max;
+        settings_.hsv_slice_h_red1_min = config.hsv_slice_h_red1_min;
+        settings_.hsv_slice_h_red1_max = config.hsv_slice_h_red1_max;
+        settings_.hsv_slice_h_red2_min = config.hsv_slice_h_red2_min;
+        settings_.hsv_slice_h_red2_max = config.hsv_slice_h_red2_max;
+        settings_.hsv_slice_s_min = config.hsv_slice_s_min;
+        settings_.hsv_slice_s_max = config.hsv_slice_s_max;
+        settings_.hsv_slice_v_min = config.hsv_slice_v_min;
+        settings_.hsv_slice_v_max = config.hsv_slice_v_max;
+
+        settings_.min_roomba_blob_size = config.min_roomba_blob_size;
+        settings_.max_roomba_blob_size = config.max_roomba_blob_size;
+
+        settings_.morphology_size = config.morphology_size;
+        settings_.morphology_iterations = config.morphology_iterations;
+    }
+}
+
+RoombaEstimatorSettings RoombaEstimator::getSettings(
+        const ros::NodeHandle& private_nh)
+{
+    RoombaEstimatorSettings settings;
+    ROS_ASSERT(private_nh.getParam(
```

Comment:
These all could be replaced easily with a macro. Might be a good place for it.

Reply:
Will do, couldn't find an easy c++ way to do this

Reply:
Fixed